app/controllers/sign_up/passwords_controller.rb

```diff
@@ -17,6 +31,51 @@ def create
 
     private
 
+    def process_successful_confirmation
+      if !@user.confirmed?
+        process_valid_confirmation_token
+      else
+        process_confirmed_user
+      end
+    end
+
+    def process_valid_confirmation_token
+      @confirmation_token = params[:confirmation_token]
+      flash.now[:notice] = t('devise.confirmations.confirmed_but_must_set_password')
+      render :new
+    end
+
+    def process_confirmed_user
+      create_user_event(:email_changed, @user)
+
+      flash[:notice] = t('devise.confirmations.confirmed')
+      redirect_to after_confirmation_path_for(@user)
+      EmailNotifier.new(@user).send_email_changed_email
+    end
+
+    def process_unsuccessful_confirmation
+      return process_already_confirmed_user if @user.confirmed?
+
+      @confirmation_token = params[:confirmation_token]
+      flash.now[:error] = unsuccessful_confirmation_error
+      render '/sign_up/email_resend/new'
+    end
+
+    def process_already_confirmed_user
+      action_text = 'Please sign in.' unless user_signed_in?
```

Comment:
I know it didn't change in this PR, but looks like a good spot for some i18n action